pytorch_lightning/core/lightning.py

```diff
@@ -1169,16 +1169,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with ``untoggle_optimizer`` to make sure param_requires_grad_state is properly reset.
+
         Args:
-            optimizer:
-            optimizer_idx:
+            optimizer: Current optimizer used in training_loop
+            optimizer_idx: Current optimizer idx in training_loop
         """
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        # make sure current optimizer is latest to be iterated over.
+        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
+        num_optimizers = len(optimizers) - 1
+        for opt_idx, opt in enumerate(optimizers):
+            for group in opt.param_groups:
+                for param in group['params']:
+                    if num_optimizers == opt_idx:
```

Comment:
can we include some early break, such as you do not need to continue in the loop after the assignment?

Reply:
I don't understand what you mean :)

Reply:
Don't think it's possible to break out the loop early, this function is ensuring that if `require_grads=False`, that we keep the original state of the parameter before setting `require_grads` False to revert the change.

@tchaton let's say I want to turn off my two optimizers for the same parameters
```python
toggle_optimizer(adam_optimizer, ...)
toggle_optimizer(sgd_optimizer, ...)
```
Will this work? I think as it currently stands the function will leave the shared parameters `require_grad=True` right?

Reply:
It should be

```
toggle_optimizer(adam_optimizer, ...)
untoggle_optimizer(adam_optimizer, ...)
toggle_optimizer(sgd_optimizer, ...)
```